src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/ReferenceCleanupService.ReferenceHandlers.cs

```diff
@@ -0,0 +1,185 @@
+                        {
+                            await access.CheckoutAsync(projectPath);
+
+                            item.SetMetadataValue("TreatAsUsed", newValue);
+
+                            wasUpdated = true;
+                        }
+                    }
+                    catch
+                    {
+                        wasUpdated = false;
+                    }
+                }, cancellationToken);
+
+                return wasUpdated;
+            }
+        }
+
+        private class ProjectReferenceHandler : ReferenceHandler
+        {
+            internal ProjectReferenceHandler() : base(ProjectSystemReferenceType.Project, ProjectReference.SchemaName, ProjectReference.SchemaName, "Identity")
+            { }
+
+            public override async Task<bool> RemoveReferenceAsync(ConfiguredProject configuredProject, ProjectSystemReferenceInfo reference)
+            {
+                Assumes.Present(configuredProject);
+                Assumes.Present(configuredProject.Services);
+                Assumes.Present(configuredProject.Services.ProjectReferences);
+
+                await configuredProject.Services.ProjectReferences.RemoveAsync(reference.ItemSpecification);
+
+                return true;
+            }
+        }
+
+        private class PackageReferenceHandler : ReferenceHandler
+        {
+            internal PackageReferenceHandler() : base(ProjectSystemReferenceType.Package, PackageReference.SchemaName, PackageReference.SchemaName, "Name")
+            { }
+
+            public override async Task<bool> RemoveReferenceAsync(ConfiguredProject configuredProject, ProjectSystemReferenceInfo reference)
+            {
+                Assumes.Present(configuredProject);
+                Assumes.Present(configuredProject.Services);
+                Assumes.Present(configuredProject.Services.PackageReferences);
+
+                await configuredProject.Services.PackageReferences.RemoveAsync(reference.ItemSpecification);
+
+                return true;
+            }
+        }
+
+        private class AssemblyReferenceHandler : ReferenceHandler
+        {
+            internal AssemblyReferenceHandler() : base(ProjectSystemReferenceType.Assembly, AssemblyReference.SchemaName, "Reference", "Identity")
+            { }
+
+            public override async Task<bool> RemoveReferenceAsync(ConfiguredProject configuredProject, ProjectSystemReferenceInfo reference)
+            {
+                Assumes.Present(configuredProject);
+                Assumes.Present(configuredProject.Services);
+                Assumes.Present(configuredProject.Services.AssemblyReferences);
+
+                AssemblyName assemblyName = new AssemblyName(reference.ItemSpecification);
+
+                await configuredProject.Services.AssemblyReferences.RemoveAsync(assemblyName, null);
+
+                return true;
+            }
+        }
+
+        private class SdkReferenceHandler : ReferenceHandler
```

Comment:
I don't think SdkReferenceHandler should exist unless Roslyn specifically handles these reference items, I not sure why we're falling back to it.

Reply:
removed